ewcasino.py

```diff
@@ -1314,6 +1319,234 @@
+
+	if challenger.poi != challengee.poi:
+		response = "Both duelers must be in the same location."
+		return await ewutils.send_message(cmd.client, cmd.message.channel, ewutils.formatMessage(author, response))
+
+	# Players have to be enlisted
+	playable_life_states = [ewcfg.life_state_enlisted, ewcfg.life_state_lucky, ewcfg.life_state_executive]
+	if challenger.life_state not in playable_life_states or challengee.life_state not in playable_life_states:
+		if challenger.life_state == ewcfg.life_state_corpse:
+			response = "Ghosts can't duel people. Alas, they can only watch from afar and !haunt."
+			return await ewutils.send_message(cmd.client, cmd.message.channel, ewutils.formatMessage(author, response))
+
+		elif challengee.life_state == ewcfg.life_state_corpse:
+			response = "Ghosts can't duel people. Alas, they can only watch from afar and !haunt."
+			return await ewutils.send_message(cmd.client, cmd.message.channel, ewutils.formatMessage(author, response))
+
+		else:
+			response = "Juveniles are too cowardly to throw their lives away in a duel."
+			return await ewutils.send_message(cmd.client, cmd.message.channel, ewutils.formatMessage(author, response))
+
+	# Assign a challenger so players can't be challenged
+	challenger.rr_challenger = challenger.id_user
+	challengee.rr_challenger = challenger.id_user
+	
+	# Apply restrictions to prevent movement, !tp, etc. This also lets the !accept command differentiate RR from duels.
+	challenger.rr_restriction = 2
+	challengee.rr_restriction = 2
+
+	challenger.persist()
+	challengee.persist()
+
+	response = "You have been challenged by {} to a duel. Do you !accept or !refuse?".format(author.display_name).replace("@", "\{at\}")
+	await ewutils.send_message(cmd.client, cmd.message.channel, ewutils.formatMessage(member, response))
+
+	# Wait for an answer
+	accepted = 0
+	try:
+		msg = await cmd.client.wait_for_message(timeout=30, author=member, check=ewutils.check_accept_or_refuse)
+
+		if msg != None:
+			if msg.content == "!accept":
+				accepted = 1
+	except:
+		accepted = 0
+
+	# Start game
+	if accepted == 1:
+		challenger = EwUser(member=author)
+		challengee = EwUser(member=member)
+
+		challenger.persist()
+		challengee.persist()
+		
+		# start the countdown. if either user tries anything funny before the countdown is over, blow their heads clean off via BHB
+		countdown = 10
+		
+		while countdown != 0:
+			text_mod = ""
+			
+			if countdown > 8:
+				text_mod = ""
+			elif countdown > 6:
+				text_mod = "*"
+			elif countdown > 4:
+				text_mod = "**"
+			else:
+				text_mod = "***"
+			
+			if countdown == 10: 
+				countdown_resp = "ENDLESS WAR begins to count down...\n{}{}{}".format(text_mod, countdown, text_mod)
+			else:
+				countdown_resp = "{}{}{}".format(text_mod, countdown, text_mod)
+				
+			countdown -= 1
+			await ewutils.send_message(cmd.client, cmd.message.channel, countdown_resp)
+
+			try:
+				msg = await cmd.client.wait_for_message(timeout=1, author=author)
```

Comment:
this should probably be done in the general on_message. it's less reliable this way. if player 2 sends a message while the bot is waiting for a response from player 1 they won't be punished.

Reply:
thinking about this, it also occurred to me that rr_challenger and rr_restriction should better be stored in a runtime variable, like with ewmap.active_moves. otherwise they might not be cleared in case the bot crashes in the middle of a duel

Reply:
also we need to reload the users' data after this await call